dsgrid/query/models.py

```diff
@@ -213,40 +205,119 @@ def handle_dimension_filters(cls, dimension_filters):
-        for agg in aggregations:
-            for dim_type in DimensionType:
-                field = dim_type.value
-                for item in getattr(agg.dimensions, field):
-                    # TODO: we could add support for this, but it does add some ambiguity.
-                    if item.function is not None:
-                        raise ValueError(
-                            f"function={item.function} cannot be set in ProjectQueryDatasetParamsModel"
-                        )
-                    if item.alias is not None:
-                        # TODO: need to support aliases here. The code that handles columns during
-                        # dataset concatenation doesn't support it.
-                        raise ValueError(
-                            f"alias={item.alias} cannot be set in ProjectQueryDatasetParamsModel"
-                        )
-        return aggregations
+
+class DatasetModel(DSGBaseModel):
+    """Specifies the datasets to use in a project query."""
+
+    dataset_id: str = Field(description="Identifier for the resulting dataset")
+    source_datasets: list[Any] = Field(
+        description="Datasets from which to read. Each must be of type DatasetBaseModel."
+    )
+    expression: str | None = Field(
+        description="Expression to combine datasets. Default is to take a union of all datasets.",
+        default=None,
+    )
+    params: ProjectQueryDatasetParamsModel = Field(
+        description="Parameters affecting datasets. Used for caching intermediate tables.",
+        default=ProjectQueryDatasetParamsModel(),
+    )
+
+    @validator("source_datasets", each_item=True)
+    def check_component(cls, component):
+        if isinstance(component, DatasetBaseModel):
+            return component
+        if component["dataset_type"] == DatasetType.STANDALONE.value:
+            return StandaloneDatasetModel(**component)
+        elif component["dataset_type"] == DatasetType.EXPONENTIAL_GROWTH.value:
+            return ExponentialGrowthDatasetModel(**component)
+        raise ValueError(f"dataset_type={component['dataset_type']} isn't supported")
+
+    @validator("expression")
+    def handle_expression(cls, expression, values):
+        if expression is None:
+            expression = " | ".join((x.dataset_id for x in values["source_datasets"]))
+        return expression
+
+
+class DatasetType(enum.Enum):
+
+    EXPONENTIAL_GROWTH = "exponential_growth"
+    STANDALONE = "standalone"
+    DERIVED = "derived"
+
+
+class DatasetBaseModel(DSGBaseModel, abc.ABC):
+    @abc.abstractmethod
+    def get_dataset_id(self) -> str:
+        """Return the primary dataset ID.
+
+        Returns
+        -------
+        str
+        """
+
+
+class StandaloneDatasetModel(DatasetBaseModel):
+
+    dataset_id: str = Field(description="Dataset identifier")
+    dataset_type: DatasetType = Field(
+        description="Type of dataset specified in a query",
+        default=DatasetType.STANDALONE,
```

Comment:
Specifying the default value here feels weird to me. But maybe it's just how it's done

Reply:
Got a better suggestion?
1. The value `standalone` needs to be in the JSON so that we know how to deserialize.
2. We don't want to make the Python user type `StandaloneDatasetModel(dataset_type=DatasetType.STANDALONE)`

Reply:
No